File: modules/graphics/src/main/java/com/wiranoid/fuzzy/graphics/Camera.java
package com.wiranoid.fuzzy.graphics;

import org.joml.Matrix4f;
import org.joml.Vector3f;

/**
 * Camera class that processes input and calculates the corresponding Eular Angles,
 * Vectors and Matrices for use in OpenGL.
 */
public class Camera {
    /**
     * Defines several possible options for camera movement.
     * Used as abstraction to stay away from window-system specific input methods.
     */
    public enum Movement {
        FORWARD,
        BACKWARD,
        LEFT,
        RIGHT
    }

    /**
     * Camera attributes
     */
    private Vector3f position;
    private Vector3f direction;
    private Vector3f up;
    private Vector3f right;
    private Vector3f worldUp;

    /**
     * Eular angles
     */
    private float yaw;
    private float pitch;

    /**
     * Camera options
     */
    private float movementSpeed;
    private float mouseSensitivity;
    private float zoom;

    public Vector3f getPosition() {
        return position;
    }

    public Vector3f getDirection() {
        return direction;
    }

    public Vector3f getUp() {
        return up;
    }

    public Vector3f getRight() {
        return right;
    }

    public Vector3f getWorldUp() {
        return worldUp;
    }

    public float getYaw() {
        return yaw;
    }

    public float getPitch() {
        return pitch;
    }

    public float getMovementSpeed() {
        return movementSpeed;
    }

    public float getMouseSensitivity() {
        return mouseSensitivity;
    }

    public float getZoom() {
        return zoom;
    }

    public Camera(Vector3f position,
                  Vector3f direction,
                  Vector3f worldUp,
                  float yaw,
                  float pitch,
                  float movementSpeed,
                  float mouseSensitivity,
                  float zoom) {
        this.position = position;
        this.direction = direction;
        this.worldUp = worldUp;
        this.yaw = yaw;
        this.pitch = pitch;
        this.movementSpeed = movementSpeed;
        this.mouseSensitivity = mouseSensitivity;
        this.zoom = zoom;
        this.updateCameraVectors();
    }

    /**
     * Returns the view matrix calculated using Eular Angles and the LookAt Matrix.
     * @return
     */
    public Matrix4f getViewMatrix() {
        return new Matrix4f().lookAt(this.position, this.position.add(this.direction, new Vector3f()), this.up);
    }

    /**
     * Processes input received from any keyboard-like input system.
     * Accepts input parameter in the form of camera defined enum (to abstract it from windowing systems)
     * @param direction
     * @param deltaTime
     */
    public void processKeyboard(Camera.Movement direction, float deltaTime) {
        float velocity = this.movementSpeed * deltaTime;
        switch (direction) {
            case FORWARD:
                this.position.add(this.direction.mul(velocity, new Vector3f()));
                break;
            case BACKWARD:
                this.position.sub(this.direction.mul(velocity, new Vector3f()));
                break;
            case LEFT:
                this.position.sub(this.right.mul(velocity, new Vector3f()));
                break;
            case RIGHT:
                this.position.add(this.right.mul(velocity, new Vector3f()));
                break;
        }
    }

    /**
     * Processes input received from a mouse input system.
     * Expects the offset value in both the x and y direction.
     * @param xOffset
     * @param yOffset
     * @param constrainPitch
     */
    public void processMouseMovement(float xOffset, float yOffset, boolean constrainPitch) {
        xOffset *= this.mouseSensitivity;
        yOffset *= this.mouseSensitivity;

        this.yaw += xOffset;
        this.pitch += yOffset;

        // Make sure that when pitch is out of bounds, screen doesn't get flipped
        if (constrainPitch) {
            if (this.pitch > 89.0f) {
                this.pitch = 89.0f;
            } else if (this.pitch < -89.0f) {
                this.pitch = -89.0f;
            }
        }

        // Update Front, Right and Up Vectors using the updated Eular angles
        this.updateCameraVectors();
    }

    /**
     * Processes input received from a mouse scroll-wheel event.
     * Only requires input on the vertical wheel-axis.
     * @param yOffset
     */
    public void processMouseScroll(float yOffset) {
        if (1.0f <= this.zoom && this.zoom <= 80.0f) {
            this.zoom -= yOffset;
        }
        if (this.zoom <= 1.0f) {
            this.zoom = 1.0f;
        }
        if (this.zoom >= 80.0f) {
            this.zoom = 80.0f;
        }
    }

    /**
     * Calculates the direction vector from the Camera's (updated) Eular Angles.
     */
    private void updateCameraVectors() {
        // Calculate the new direction vector
        Vector3f direction = new Vector3f();
        direction.x = (float) ( Math.cos(Math.toRadians(this.yaw)) * Math.cos(Math.toRadians(this.pitch)) );
        direction.y = (float) Math.sin(Math.toRadians(this.pitch));
        direction.z = (float) ( Math.sin(Math.toRadians(this.yaw)) * Math.cos(Math.toRadians(this.pitch)) );
        this.direction = direction.normalize();
        // Also recalculate the Right and Up vector
        this.right = this.direction.cross(this.worldUp, new Vector3f()).normalize();
        this.up = this.right.cross(this.direction, new Vector3f()).normalize();
    }

}

File: modules/graphics/src/Camera.cpp
#include "../include/Camera.h"

#include <GL/glew.h>
#include <glm/glm.hpp>
#include <glm/gtc/matrix_transform.hpp>
#include <math.h>

using namespace graphics;

Camera::Camera(glm::vec3 position, glm::vec3 up, GLfloat yaw, GLfloat pitch) :
        Position(position),
        WorldUp(up),
        Yaw(yaw),
        Pitch(pitch),
        Front(glm::vec3(0.0f, 0.0f, -1.0f)),
        MovementSpeed(3.0f),
        MouseSensitivity(0.1f),
        Zoom(45.0f) {
    this->updateCameraVectors();
}

Camera::Camera(GLfloat posX,
               GLfloat posY,
               GLfloat posZ,
               GLfloat upX,
               GLfloat upY,
               GLfloat upZ,
               GLfloat yaw,
               GLfloat pitch) :
        Yaw(yaw),
        Pitch(pitch),
        Front(glm::vec3(0.0f, 0.0f, -1.0f)),
        MovementSpeed(3.0f),
        MouseSensitivity(0.25f),
        Zoom(45.0f) {
    this->Position = glm::vec3(posX, posY, posZ);
    this->WorldUp = glm::vec3(upX, upY, upZ);
    this->updateCameraVectors();
}

glm::mat4 Camera::GetViewMatrix() {
    return glm::lookAt(this->Position, this->Position + this->Front, this->Up);
}

void Camera::ProcessKeyboard(Camera::Movement direction, GLfloat deltaTime) {
    GLfloat velocity = this->MovementSpeed * deltaTime;
    if (direction == FORWARD)
        this->Position += this->Front * velocity;
    if (direction == BACKWARD)
        this->Position -= this->Front * velocity;
    if (direction == LEFT)
        this->Position -= this->Right * velocity;
    if (direction == RIGHT)
        this->Position += this->Right * velocity;
}

void Camera::ProcessMouseMovement(GLfloat xoffset, GLfloat yoffset, GLboolean constrainPitch) {
    xoffset *= this->MouseSensitivity;
    yoffset *= this->MouseSensitivity;

    this->Yaw   += xoffset;
    this->Pitch += yoffset;

    // Make sure that when pitch is out of bounds, screen doesn't get flipped
    if (constrainPitch)
    {
        if (this->Pitch > 89.0f)
            this->Pitch = 89.0f;
        if (this->Pitch < -89.0f)
            this->Pitch = -89.0f;
    }

    // Update Front, Right and Up Vectors using the updated Eular angles
    this->updateCameraVectors();
}

void Camera::ProcessMouseScroll(GLfloat yoffset) {
    if (this->Zoom >= 1.0f && this->Zoom <= 45.0f)
        this->Zoom -= yoffset;
    if (this->Zoom <= 1.0f)
        this->Zoom = 1.0f;
    if (this->Zoom >= 45.0f)
        this->Zoom = 45.0f;
}

void Camera::updateCameraVectors() {
    // Calculate the new Front vector
    glm::vec3 front;
    front.x = cos(glm::radians(this->Yaw)) * cos(glm::radians(this->Pitch));
    front.y = sin(glm::radians(this->Pitch));
    front.z = sin(glm::radians(this->Yaw)) * cos(glm::radians(this->Pitch));
    this->Front = glm::normalize(front);
    // Also re-calculate the Right and Up vector.
    // Normalize the vectors, because their length gets closer to 0
    // the more you look up or down which results in slower movement.
    this->Right = glm::normalize(glm::cross(this->Front, this->WorldUp));
    this->Up    = glm::normalize(glm::cross(this->Right, this->Front));
}
